plugins/inputs/dpdk/dpdk.go

```diff
@@ -0,0 +1,240 @@
+		}
+		return dpdk
+	})
+}
+
+func (dpdk *dpdk) SampleConfig() string {
+	return dpdkSampleConfig
+}
+
+func (dpdk *dpdk) Description() string {
+	return dpdkDescription
+}
+
+// Performs validation of all parameters from configuration
+func (dpdk *dpdk) Init() error {
+	if dpdk.SocketPath == "" {
+		dpdk.SocketPath = defaultPathToSocket
+		dpdk.Log.Infof("using default '%v' path for socket_path", defaultPathToSocket)
+	}
+	if err := isSocket(dpdk.SocketPath); err != nil {
+		return err
+	}
+
+	if err := dpdk.validateCommands(); err != nil {
+		return err
+	}
+
+	if dpdk.AccessTimeout < 0 {
+		return fmt.Errorf("socket_access_timeout should be positive number or equal to 0 (to disable timeouts)")
+	}
+	if len(dpdk.AdditionalCommands) == 0 && len(dpdk.DeviceTypes) == 0 {
+		dpdk.Log.Warn("DPDK plugin is enabled and configured with nothing to read")
+	}
+
+	dpdk.connector = newDpdkConnector(dpdk.SocketPath, dpdk.AccessTimeout)
+	result, err := dpdk.connector.connect()
+	if result != "" {
+		dpdk.Log.Info(result)
+	}
+	return err
+}
+
+// Checks that user-supplied commands are unique and match DPDK commands format
+func (dpdk *dpdk) validateCommands() error {
+	dpdk.AdditionalCommands = uniqueValues(dpdk.AdditionalCommands)
+
+	for _, commandWithParams := range dpdk.AdditionalCommands {
+		if len(commandWithParams) == 0 {
+			return fmt.Errorf("got empty command")
+		}
+
+		if commandWithParams[0] != '/' {
+			return fmt.Errorf("'%v' command should start with '/'", commandWithParams)
+		}
+
+		if commandWithoutParams := stripParams(commandWithParams); len(commandWithoutParams) >= maxCommandLength {
+			return fmt.Errorf("'%v' command is too long. It shall be less than %v characters", commandWithoutParams, maxCommandLength)
+		}
+
+		if len(commandWithParams) >= maxCommandLengthWithParams {
+			return fmt.Errorf("command with parameters '%v' shall be less than %v characters", commandWithParams, maxCommandLengthWithParams)
+		}
+	}
+
+	return nil
+}
+
+// Gathers all unique commands and processes each command sequentially
+// Parallel processing could be achieved by running several instances of this plugin with different settings
+func (dpdk *dpdk) Gather(accumulator telegraf.Accumulator) error {
+	commands := dpdk.gatherCommands(accumulator)
+
+	for _, command := range commands {
+		dpdk.processCommand(command, accumulator)
+	}
+
+	return nil
+}
+
+// Gathers all unique commands
+func (dpdk *dpdk) gatherCommands(accumulator telegraf.Accumulator) []string {
+	var commands []string
+	if contains(dpdk.DeviceTypes, "ethdev") {
+		commands = append(commands, dpdk.getCommandsAndParamsCombinations(dpdk.ethdevCommands, dpdk.EthdevConfig.EthdevExcludeCommands, ethdevListCommand, accumulator)...)
+	}
+
+	if contains(dpdk.DeviceTypes, "rawdev") {
+		commands = append(commands, dpdk.getCommandsAndParamsCombinations(dpdk.rawdevCommands, []string{}, rawdevListCommand, accumulator)...)
```

Comment:
I think you should only pass `acc` to this function as everything else is also reachable from within the function.

Reply:
@srebhan 
I agree that everything except `acc` is also reachable from within `dpdk.getCommandsAndParamsCombinations` function but this was done intentionally -> to not double operations which are done inside `dpdk.getCommandsAndParamsCombinations` function. I would leave it as it is.

However, if you see how these two functions can be improved, I would be more than happy to get meaningful example.

```
func (dpdk *dpdk) gatherCommands(accumulator telegraf.Accumulator) []string {
	var commands []string
	if contains(dpdk.DeviceTypes, "ethdev") {
		commands = append(commands, dpdk.getCommandsAndParamsCombinations(dpdk.ethdevCommands, dpdk.EthdevConfig.EthdevExcludeCommands, ethdevListCommand, accumulator)...)
	}

	if contains(dpdk.DeviceTypes, "rawdev") {
		commands = append(commands, dpdk.getCommandsAndParamsCombinations(dpdk.rawdevCommands, []string{}, rawdevListCommand, accumulator)...)
	}

	commands = append(commands, dpdk.AdditionalCommands...)

	return uniqueValues(commands)
}

func (dpdk *dpdk) getCommandsAndParamsCombinations(appendedCommands []string, excludedCommands []string, listCommand string, accumulator telegraf.Accumulator) []string {
	validCommands := removeSubset(appendedCommands, excludedCommands)

	commands, err := dpdk.appendCommandsWithParamsFromList(listCommand, validCommands)
	if err != nil {
		accumulator.AddError(fmt.Errorf("error occurred during fetching of %v params - %v", listCommand, err))
	}
	return commands
}
```

Reply:
Then why not fuse the two an duplicate the two lines?

Reply:
Fused as suggested.